pandas/core/indexes/base.py

```diff
@@ -3111,46 +3111,125 @@ def join(self, other, how='left', level=None, return_indexers=False,
         if not len(overlap):
-            raise ValueError("cannot join with no level specified and no "
-                             "overlapping names")
-        if len(overlap) > 1:
-            raise NotImplementedError("merging with more than one level "
-                                      "overlap on a multi-index is not "
-                                      "implemented")
-        jl = overlap[0]
+            raise ValueError("cannot join with no overlapping index names")
+
+        # Drop the non matching levels
+        ldrop_levels = list(set(self_names) - overlap)
+        rdrop_levels = list(set(other_names) - overlap)
+
+        self_is_mi = isinstance(self, MultiIndex)
+        other_is_mi = isinstance(other, MultiIndex)
+
+        if self_is_mi and other_is_mi:
+            self_tmp = self.droplevel(ldrop_levels)
+            other_tmp = other.droplevel(rdrop_levels)
+
+            if not (self_tmp.is_unique and other_tmp.is_unique):
+                raise ValueError("Join on level between two MultiIndex objects"
+                                 "is ambiguous")
+
+            # Append to the returned Index the non-overlapping levels
+            dropped_lvls = ldrop_levels + rdrop_levels
+
+            join_index, lidx, ridx = self_tmp.join(other_tmp, how=how,
+                                                   return_indexers=True)
+
+            if how == 'outer':
+                levels, labels, names = self._complete_outer_join(other,
+                                                                  join_index,
+                                                                  lidx, ridx,
+                                                                  dropped_lvls)
+
+            else:
+                if how == 'left':
+                    join_index = self
+                elif how == 'right':
+                    join_index = other
+
+                levels = join_index.levels
+                labels = join_index.labels
+                names = join_index.names
+
+            join_index = MultiIndex(levels=levels, labels=labels,
+                                    names=names, verify_integrity=False)
+
+            return join_index, lidx, ridx
+
+        jl = list(overlap)[0]
 
         # make the indices into mi's that match
-        if not (self_is_mi and other_is_mi):
-
-            flip_order = False
-            if self_is_mi:
-                self, other = other, self
-                flip_order = True
-                # flip if join method is right or left
-                how = {'right': 'left', 'left': 'right'}.get(how, how)
-
-            level = other.names.index(jl)
-            result = self._join_level(other, level, how=how,
-                                      return_indexers=return_indexers)
-
-            if flip_order:
-                if isinstance(result, tuple):
-                    return result[0], result[2], result[1]
-            return result
+        flip_order = False
```

Comment:
huh?

if you need to do something based on the how, then do it above

Reply:
The part from `jl = list(overlap)[0]` and onwards refers to cases where not both indexes are multiindexes.
This is legacy code that I didn't modify. Shall I try to rewrite?